sn_client/src/audit/mod.rs

```diff
@@ -0,0 +1,251 @@
+                );
+                trace!("Spends for {parent_tx_hash:?} - {spends:?}");
+
+                // check if we reached the genesis Tx
+                if parent_tx == sn_transfers::GENESIS_CASHNOTE.src_tx
+                    && spends
+                        .iter()
+                        .all(|s| s.spend.unique_pubkey == sn_transfers::GENESIS_CASHNOTE.id)
+                    && spends.len() == 1
+                {
+                    debug!("Depth {depth} - Reached genesis Tx on one branch: {parent_tx_hash:?}");
+                    verified_tx.insert(parent_tx_hash);
+                    continue;
+                }
+
+                // verify tx with those spends
+                parent_tx
+                    .verify_against_inputs_spent(&spends)
+                    .map_err(|err| WalletError::CouldNotVerifyTransfer(format!("at depth {depth} - Failed to verify parent Tx {parent_tx_hash:?}: {err}")))?;
+                verified_tx.insert(parent_tx_hash);
+                debug!("Depth {depth} - Verified parent Tx: {parent_tx_hash:?}");
+
+                // add new parent spends to next gen
+                next_gen_tx.extend(spends.into_iter().map(|s| s.spend.parent_tx));
+            }
+
+            // only verify parents we haven't already verified
+            txs_to_verify = next_gen_tx
+                .into_iter()
+                .filter(|tx| !verified_tx.contains(&tx.hash()))
+                .collect();
+
+            depth += 1;
+            let elapsed = start.elapsed();
+            let n = verified_tx.len();
+            println!("Now at depth {depth} - Verified {n} transactions in {elapsed:?}");
+        }
+
+        let elapsed = start.elapsed();
+        let n = verified_tx.len();
+        println!("Verified all the way to genesis! Through {depth} generations, verifying {n} transactions in {elapsed:?}");
+        Ok(())
+    }
+
+    /// This function does the opposite of verify_spend.
+    /// It recursively follows the descendants of a Spend, all the way to unspent Transaction Outputs (UTXOs).
+    ///
+    /// Prints progress on stdout
+    ///
+    /// Starting from Genesis, this amounts to Auditing the entire currency.
+    /// This is how the DAG it follows could look like:
+    ///
+    /// ```text
+    ///                                   -> Spend7 ---> UTXO_11
+    ///                                 /
+    /// Genesis -> Spend1 -----> Spend2 ---> Spend5 ---> UTXO_10
+    ///                   \
+    ///                     ---> Spend3 ---> Spend6 ---> UTXO_9
+    ///                     \
+    ///                       -> Spend4 ---> UTXO_8
+    ///
+    /// gen0       gen1          gen2        gen3
+    ///
+    /// ```
+    ///
+    /// This function will return the UTXOs (Spend addresses not spent yet)
+    /// Future calls to this function could start from those UTXOs to avoid
+    /// re-checking all previously checked branches.
+    pub async fn follow_spend(
```

Comment:
this is following a single spend? Or it's doing the whole token dag? name doesnt fit, I think if we're doing the whole dag